src/server/transaction.h

```diff
@@ -552,17 +553,20 @@ class Transaction {
   }
 
  private:
-  // shard_data spans all the shards in ess_.
-  // I wish we could use a dense array of size [0..uniq_shards] but since
-  // multiple threads access this array to synchronize between themselves using
-  // PerShardData.state, it can be tricky. The complication comes from multi_ transactions where
-  // scheduled transaction is accessed between operations as well.
-
-  // Stores per-shard data.
-  // For non-multi transactions, it can be of size one in case only one shard is active
-  // (unique_shard_cnt_ = 1).
-  // Never access directly with index, always use SidToId.
-  absl::InlinedVector<PerShardData, 4> shard_data_;  // length = shard_count
+  // Main synchronization point for dispatching hop callbacks and waiting for them to finish.
+  // After scheduling, sequential hops are executed as follows:
+  //    coordinator: Prepare hop, then Start(num_shards), dispatch poll jobs and Wait()
+  //    tx queue:    Once IsArmedInShard() /* checks Active() */ -> run in shard and Dec()
+  // As long as barrier is active any writes by the coordinator are prohibited, so shard threads
+  // can safely read transaction state and modify per-shard state belonging to them.
+  // Inter-thread synchronization is provided by the barriers acquire/release pairs.
+  PhasedBarrier run_barrier_;
```

Comment:
Now our inter-thread coordination during hops is **clearly** defined in a single place in simple terms